### ISSUE TITLE
canary-1b-flash not generalizing

### ISSUE THREAD

**Issue by amarbayar** [NVIDIA/NeMo#12845]
**Describe the bug**

Thanks for making `canary-1b-flash` available to fine-tune! We have tried fine-tuning it for a new language and got it down to 8% WER which was exciting to see. Evaluation was being done only on the eval dataset. When we tested the model against generic data the model performed extremely poorly to the point it made us wonder how it reached 8% WER -- and that of course is a classic sign of overfitting. Given there is no official documentation/guidance on fine-tuning `canary-1b-flash` or `canary-1b`, we went through the code to figure out how to do it so it is possible we did some mistakes on the fine-tuning side (params, training script, dataset prep). So, wanted to ask if what we have used looks fine and if so how the model could be overfitting easily and not able to generalize well on unseen dataset?

**Steps/Code to reproduce bug**

1. Prepared the dataset including (train manifest and eval manifest) 
2. Enabled lhotse
3. Created tokenizer model, vocab and tokenizer vocab
4. Created special tokens for the new language and updated the config
5. Updated the base pre-trained model with the config with spl token
6. The following is the script that leverages `speech_to_text_aed.py` script:```python /workspace/nemo/examples/asr/speech_multitask/speech_to_text_aed.py \--config-path=/workspace/nemo/examples/asr/conf/speech_multitask \
--config-name=fast-conformer_aed.yaml \
model.train_ds.manifest_filepath=/workspace/nemo/dataset/tm.json \
model.train_ds.use_lhotse=true \
model.train_ds.batch_size=8 \
model.train_ds.shuffle=true \
model.train_ds.sample_rate=24000 \
model.train_ds.num_workers=8 \
model.train_ds.batch_duration=360 \
+model.train_ds.text_field=text \
+model.train_ds.lang_field=target_lang \
++model.train_ds.num_buckets=20 \
++model.train_ds.pretokenize=false \
model.validation_ds.manifest_filepath=/workspace/nemo/dataset/vm.json \
model.validation_ds.use_lhotse=true \
model.validation_ds.batch_size=8 \
model.validation_ds.shuffle=false \
model.validation_ds.sample_rate=24000 \
model.validation_ds.num_workers=4 \
+model.validation_ds.text_field=text \
+model.validation_ds.lang_field=target_lang \
++model.validation_ds.num_buckets=20 \
++model.validation_ds.pretokenize=false \
model.test_ds=null \
model.prompt_format=canary \
model.decoding.strategy=greedy \
trainer.devices=1 \
trainer.precision=16 \
trainer.max_epochs=10 \
trainer.accumulate_grad_batches=4 \
++trainer.val_check_interval=10000 \
+trainer.limit_val_batches=100 \
++optim.sched.warmup_steps=27000 \
+optim.lr=0.00001 \
init_from_nemo_model=/workspace/nemo/canary-1b-xx-updated.nemo \
exp_manager.resume_ignore_no_checkpoint=true \
+exp_manager.checkpoint_callback_params.every_n_epochs=1 \
+exp_manager.checkpoint_callback_params.save_last=true \
+exp_manager.checkpoint_callback_params.dirpath=/workspace/nemo/checkpoints \
+exp_manager.checkpoint_callback_params.filename='"checkpoint-{epoch}-{val_wer:.2f}"'
```

One thing we did was we sampled all of our audios to 24,000Hz. We did see notice about 16,000Hz but not specific to canary-1b{-flash}. 

Again, evaluations were almost perfect and reference and predictions indeed looked 8% or even better at times. But the model just isn't generalizing to unseen data. 

**Expected behavior**

Model generalizes to unseen dataset. 

**Environment overview (please complete the following information)**

 - Environment location: Cloud (Shared GPU)
 - Method of NeMo install: Installed from source since we had to tweak the config and invoke the aed script
 - If method of install is [Docker], provide `docker pull` & `docker run` commands used

**Environment details**

If NVIDIA docker image is used you don't need to specify these.
Otherwise, please provide:
- OS version: Ubuntu 22.04
- PyTorch version: 2.2.0
- Python version: 3.10

**Additional context**

Add any other context about the problem here.
- Tried on NVIDIA A40

**Comment by amarbayar** [comment]
@jeremy110  - You mean hours of dataset or training time? Mine is the 1b-flash.

**Comment by jeremy110** [comment]
@amarbayar 
I mean hours of training dataset.
Basically, the two models are quite similar, except that Flash has more encoder layers and has been trained with additional timestamps.

**Comment by amarbayar** [comment]
@jeremy110  - I ran mine on close to 2000 hours of dataset. Do you have your training code somewhere by any chance?

**Comment by jeremy110** [comment]
@amarbayar 
The original model used BPE tokens, but since this approach does not work well for Japanese, Korean, and Chinese, I adopted Byte-level BPE (BBPE) to reduce the vocabulary size, which significantly improved convergence speed.

Additionally, since I am using an RTX 4090 with 24GB of memory, I switched from the original AdamW optimizer to bnb AdamW8bit to reduce GPU memory usage. I also replaced nn.Embedding in the decoder with bnb.nn.StableEmbedding.

These are my optimizations.

Furthermore, I discovered some bugs in change_vocabulary:

1. The tokenizer_cfg needs to be reassigned; otherwise, loading the trained model will cause issues.

2. self.wer needs to be updated; otherwise, WER calculations will be incorrect.

```python
# Setup the tokenizer
self._setup_tokenizer(tokenizer_cfg)
self.cfg.tokenizer = tokenizer_cfg  # [ADD]

self.decoding = MultiTaskDecoding(
    decoding_cfg=decoding_cfg,
    transformer_decoder=self.transf_decoder,
    log_softmax_module=self.log_softmax,
    tokenizer=self.tokenizer,
)
self.wer = WER(self.decoding, log_prediction=self.cfg.get("log_prediction"))  # [ADD]
```

These are my modifications—hope they help!

**Comment by sutungpo** [comment]
> Can I ask how many hours of training you have used? Because I tried training with approximately 1,000 hours of Japanese and Korean, and the eval WER is around 25%. On the Fleurs test set, Japanese is around 13%, and Korean is around 33%. However, it works well for recognizing unseen scenarios.
> 
> My pre-trained model is canary-0.6b turbo.

I'm also interested in Japanese. Whisper-v3 achieved a 10.3% WER on the Japanese portion of the Fleurs dataset. Can this model be fine-tuned to surpass Whisper's performance?

**Comment by jeremy110** [comment]
@sutungpo 
I think it can be exceeded, but the training data needs to be more, at least 10k hours, but I don’t have so many GPUs to test it.

**Comment by ankitapasad** [comment]
Hi @amarbayar 

Thank you for your interest! 

Can you elaborate on what you mean by testing on generic data?

The question says that you get to 8% WER on eval set, which would imply that the model is not overfitting. Note that once you fine-tune the model on a new task/data/language the model may forget some of the original capabilities it is trained for. One way to circumvent this is to make sure you add some of the task/language data for the original capabilities that you wish to retain in the fine-tuned model.

Hope that helps!

**Comment by amarbayar** [comment]
@ankitapasad - Thanks for your response. 

Yeah I did get an impressive WER 8% on my dataset which was great to see. But when I tested with other data outside of my dataset the results were very poor to the point where the model was producing the repeated characters...etc. And occasionally it transcribed some words okay but majority of my test cases (outside of my dataset) performed very poor. 

Any idea how this could have happened?

**Comment by ankitapasad** [comment]
Hi @amarbayar, does this other data correspond to your finetuning language? If not, then unfortunately some degradation is expected, because after finetuning the model on 2000 hours of the new language, the model will tend to forget other languages it's already trained on. The most straightforward fix is have at least some amount of training data for other languages in the fine-tuning data to make sure that the model does not forget its original capabilities.

**Comment by amarbayar** [comment]
@ankitapasad - Indeed, same language. Dataset used for training and dataset used for testing are both exact same language. So this is not the catastrophic forgetting issue.

**Comment by Shaheer66** [comment]
@amarbayar Hi, the whole guide from your side is amazing.
Can you provide a detail guide how to finetune the nvidia/canary-1b-v2
As my targeted model is also trained on the Portguese language so guide me acordingly
Actually I am interested in Brazilian-Portguese language and It is not my native language, so I need guidance.
I have  to achieve more than 90% accuracy that means wer should be less than 10%.
If you can guide me through this, I shall be very thankful to you for this act of kindness
Looking forward to discuss more about this
